distsys/resources/tcpmailboxes.go

```diff
@@ -0,0 +1,394 @@
+package resources
+
+import (
+	"encoding/gob"
+	"fmt"
+	"github.com/UBC-NSS/pgo/distsys"
+	"io"
+	"net"
+	"time"
+)
+
+// Mailboxes as Archetype Resource
+// -------------------------------
+
+const (
+	tcpNetworkBegin = iota
+	tcpNetworkValue
+	tcpNetworkPreCommit
+	tcpNetworkCommit
+)
+
+type TCPMailboxKind int
+
+const (
+	TCPMailboxesLocal = iota
+	TCPMailboxesRemote
+)
+
+const (
+	tcpMailboxesReceiveChannelSize          = 100                   // TODO: this should be a configuration option
+	tcpMailboxesTCPTimeout                  = 1 * time.Second       // TODO: same as above
+	tcpMailboxesReadTimeout                 = 20 * time.Millisecond // TODO: same
+	tcpMailboxesConnectionDroppedRetryDelay = 50 * time.Millisecond // TODO: same
+)
+
+// AddressMappingFn is responsible for translating the index, as in network[index] from distsys.TLAValue to a pair of
+// TCPMailboxKind and address string, where the address string would be appropriate to pass to net.Listen("tcp", ...)
+// or net.Dial("tcp", ...). It should return TCPMailboxesLocal if this node is to be the only listener, and it should
+// return TCPMailboxesRemote if the mailbox is remote and should be dialed. This could potentially allow unusual setups
+// where a single process "owns" more than one mailbox.
+type AddressMappingFn func(distsys.TLAValue) (TCPMailboxKind, string)
+
+// TCPMailboxesArchetypeResourceMaker produces a distsys.ArchetypeResourceMaker for a collection of TCP mailboxes.
+// Each individual mailbox will match the following mapping macro, assuming exactly one process "reads" from it:
+//
+//     mapping macro TCPChannel {
+//       read {
+//           await Len($variable) > 0;
+//           with (msg = Head($variable)) {
+//               $variable := Tail($variable);
+//               yield msg;
+//           };
+//       }
+//
+//       write {
+//           await Len($variable) < BUFFER_SIZE;
+//           yield Append($variable, $value);
+//       }
+//     }
+//
+// As is shown above, each mailbox should be a fully reliable FIFO channel, which these resources approximated
```

Comment:
Good to mention that we don't guarantee liveness properties here.

Reply:
Good point, there are even specific lifecycle points that are not live, vs most that are.